lib/assets/javascripts/twine.coffee

```diff
@@ -46,33 +46,47 @@
     this
 
   Twine.bind = (node = rootNode, context = Twine.context(node)) ->
-    bind(context, node, true)
+    bind(context, node, getIndexesForElement(node), true)
 
   Twine.afterBound = (callback) ->
     if currentBindingCallbacks
       currentBindingCallbacks.push(callback)
     else
       callback()
 
-  bind = (context, node, forceSaveContext) ->
+  bind = (context, node, indexes, forceSaveContext) ->
     currentBindingCallbacks = []
     if node.bindingId
       Twine.unbind(node)
 
+    if defineArrayAttr = Twine.getAttribute(node, 'define-array')
+      newIndexes = defineArray(node, context, defineArrayAttr)
+      for key, value of indexes || {} when !newIndexes[key]?
+        newIndexes[key] = value
+      indexes = newIndexes
+      # register the element early because subsequent bindings on the same node might need to make use of the index
+      element = findOrCreateElementForNode(node)
+      element.indexes = indexes
+
     for type, binding of Twine.bindingTypes when definition = Twine.getAttribute(node, type)
-      element = {bindings: []} unless element  # Defer allocation to prevent GC pressure
+      element = findOrCreateElementForNode(node)
+      element.bindings ?= []
+      element.indexes ?= indexes
```

Comment:
You're sure you can use `?=` here? That checks explicitly for `null` and nothing else. If `element.indexes` is `undefined`, it will stay `undefined` in this case. Is it maybe safer to use `element.indexes ||= indexes` and cover both cases?